Fixes to Multiplayer Gem Configuration Doc (#1402)
o3de/o3de.org: content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/configuration.md

## Build setup

> As usual, the first step of enabling the Multiplayer Gem is to make sure the gem is [added to the project](/docs/user-guide/project-config/add-remove-gems/).

Reviewer: "As usual" is not adding anything here. Keep it simple, "Start of by ensuring that Multiplayer Gem is configured for the project... ] or similar.

Author: Good call! K.I.S.

After adding the Multiplayer Gem to the project there are more steps that require editing some of the project's CMake and C++ files by hand. 
### CMakeList.txt changes
Make sure the `<ProjectName>.Static` target includes the correct dependencies. 
Find the CMake file defining your project's static target; for example `<ProjectName>/Code/CMakeList.txt`. 
Make the following edits to the `<ProjectName>.Static` target:
1. In the `FILES_CMAKE` section, add `<projectname>_autogen_files.cmake`
    ```cmake
        ly_add_target(
            NAME <ProjectName>.Static STATIC
            ...
            FILES_CMAKE
                ...
                <projectname>_autogen_files.cmake
    ```

>     {{< note >}}Creating the new `<projectname>_autogen_files.cmake` file is covered below in the [Adding AutoGen CMake file](#adding_autogen_file) section.{{< /note >}}

Reviewer: "Create a new..",

Maybe "The <projectname>_autogen_files.cmake" file may not exist at this time. We will create this file later in the [Adding AutoGen CMake file](#adding_autogen_file) section. For now we just want to create reference to it in the CMakeList.txt"

Author: Fixed! Thank you

1. In the `BUILD_DEPENDENCIES PUBLIC` section, add `AZ::AzNetworking` and `Gem::Multiplayer`.
   ```cmake
    ly_add_target(
        NAME <ProjectName>.Static STATIC
        ...
        BUILD_DEPENDENCIES
            PUBLIC
                AZ::AzNetworking
                Gem::Multiplayer
   ```

>     {{< note >}}`BUILD_DEPENDENCIES` might not contain a `PUBLIC` section, add it if necessary.{{< /note >}}

Reviewer: Maybe, "If BUILD_DEPENDENCIES does not contain a public section, add it as above".

Author: I like that a lot, easier to read in an "If -> then" format. Fixed!

## Build setup

Start by ensuring the Multiplayer Gem is [added to the project](/docs/user-guide/project-config/add-remove-gems/).
After adding the Multiplayer Gem to the project there are more steps that require editing some of the project's CMake and C++ files by hand. 
### CMakeList.txt changes
Make sure the `<ProjectName>.Static` target includes the correct dependencies. 

> Find the CMake file defining your project's static target; for example `<ProjectName>/Code/CMakeList.txt`. 

Reviewer: If I generate a new project with O3DE, there is a <ProjectName>/Gem/CmakeList.txt, theres is no <ProjectName/Code/CMakeList.txt. Should that be mentioned as an example here?

Author: Aye, that's why I chose "for example" since we have different project types. There are many locations where I suggest a path (using the default path for external projects; which I'm guessing you've create a project tied to the engine?)

Make the following edits to the `<ProjectName>.Static` target:
1. In the `FILES_CMAKE` section, add `<projectname>_autogen_files.cmake`
    ```cmake
        ly_add_target(
            NAME <ProjectName>.Static STATIC
            ...
            FILES_CMAKE
                ...
                <projectname>_autogen_files.cmake
    ```

>     {{< note >}}The `<projectname>_autogen_files.cmake` file might not exist at this time. We will create this file later in the [Adding AutoGen CMake file](#adding_autogen_file) section. For now we just want to create a reference to it in the `CMakeList.txt`{{< /note >}}

Reviewer 2: ```suggestion
    {{< note >}}The `<projectname>_autogen_files.cmake` file is created later in the [Adding AutoGen CMake file](#adding_autogen_file) section. For now, create a reference to it in the `CMakeList.txt`{{< /note >}}
```

        *.AutoComponent.xml,AutoComponentTypes_Source.jinja,$path/AutoComponentTypes.cpp
)
```
<a id="adding_autogen_file"></a>
### Adding AutoGen CMake file

> Next, create a new file called `<projectname>_autogen_files.cmake` and make it a sibling of `CMakeList.txt`. Example: `<ProjectName>/Code/<projectname>_autogen_files.cmake`. The contents of this file add the source templates for [autocomponents](./autocomponents) to the project build.

Reviewer: nit: "Make a sibling of CmakeList.txt" is over complicating things. Maybe "Next, create a new file called `<projectname>_autogen_files.cmake` and place it in the Project's code folder. For example `<ProjectName>/Code/<projectname>_autogen_files.cmake`."

)
```

### Adding a temporary auto-component

> There's currently a [bug](https://github.com/o3de/o3de/issues/4058) causing a build failure if multiplayer auto-components are enabled, but no auto-components are created. As a work-around, create a temporary auto-component.

Reviewer: Move this to a {{note}}?

Author: Fixed!

1. Create a new folder under your project's `Code\Source\` directory called `AutoGen`. 

>     {{< note >}}This AutoGen directory doesn't have to be temporary; all multiplayer auto-components can live here.{{< /note >}}

Reviewer: Maybe "All future auto-components can live here"?

Author: 👍

> 1. Create a new, temporary auto-component file under `Code\Source\AutoGen` called `MyFirstNetworkComponent.AutoComponent.xml`.

Reviewer: Add note that name could be anything, its just important to ensure name is used consistently. For this guide we will use "MyFirstNetworkComponent" ?

Author: Good call out.

1. Modify `Code\Source\AutoGen\MyFirstNetworkComponent.AutoComponent.xml` to have the following content:
    ```xml  
    <?xml version="1.0"?>

    <Component
        Name="MyFirstNetworkComponent"
        Namespace="<ProjectName>"
        OverrideComponent="false"
        OverrideController="false"
        OverrideInclude=""
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    </Component>
    ```
    {{< important >}}Make sure the "Namespace" property is wrapped in quotes and matches the name of your project.
    {{< /important >}}
1. Register the temporary auto-component with CMake by updating `<projectname_files.cmake>`
    ```cmake
    set(FILES
        ...
        Source/AutoGen/MyFirstNetworkComponent.AutoComponent.xml
    )
    ```

> {{< note >}}Feel free to delete this temporary auto-component, or use it as a starting point, once you're ready to build a real multiplayer component. Find out more in general about multiplayer auto-components [here](/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autocomponents/), or follow [this tutorial](/docs/learning-guide/tutorials/multiplayer/first-multiplayer-component/).{{< /note >}}

Reviewer: nit: "once you're ready to build further multiplayer components".

Author: Yes! Thank you :)

)
```

### Adding a temporary auto-component
{{< note >}}

> There's currently a [bug](https://github.com/o3de/o3de/issues/4058) causing a build failure if multiplayer auto-components are enabled, but no auto-components are created. As a work-around, create a temporary auto-component.

Reviewer: nit: Prefer issue over bug.

Author: Fixed!

> {{< note >}}
> There's currently an [issue](https://github.com/o3de/o3de/issues/4058) causing a build failure if multiplayer auto-components are enabled, but no auto-components are created. As a work-around, create a temporary auto-component.
> {{< /note >}}

Reviewer 2: Let's reword this a bit the the probability that this note  ends up hanging around past it's useful life.

```suggestion
{{< note >}}
You might experience a build failure if multiplayer auto-components are enabled, but no auto-components are created. As a work-around, create a temporary auto-component. Refer to this [issue](https://github.com/o3de/o3de/issues/4058) for more information.
{{< /note >}}
```

1. Create a new folder under your project's `Code\Source\` directory called `AutoGen`. 
    {{< note >}}This AutoGen directory doesn't have to be temporary. All future multiplayer auto-components can live here.{{< /note >}}
1. Create a new, temporary auto-component file under `Code\Source\AutoGen` called `MyFirstNetworkComponent.AutoComponent.xml`.

>     {{< note >}}The name of this multiplayer auto-component can be anything, it's just important to ensure the name is used consistently. For this guide we will use "MyFirstNetworkComponent".{{< /note >}}

Reviewer 2: ```suggestion
    {{< note >}}This guide uses "MyFirstNetworkComponent" as the name for this multiplayer auto-component. You can specify any name for your component, but ensure that the name is used consistently. {{< /note >}}
```

1. Modify `Code\Source\AutoGen\MyFirstNetworkComponent.AutoComponent.xml` to have the following content:
    ```xml  
    <?xml version="1.0"?>

    <Component
        Name="MyFirstNetworkComponent"
        Namespace="<ProjectName>"
        OverrideComponent="false"
        OverrideController="false"
        OverrideInclude=""
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    </Component>
    ```

>     {{< important >}}Make sure the "Namespace" property is wrapped in quotes and matches the name of your project.

Reviewer 2: Minor, but this is something an editor might call out.

```suggestion
    {{< important >}}Make sure that the "Namespace" property is wrapped in quotes and matches the name of your project.
```

    {{< /important >}}
1. Register the temporary auto-component with CMake by updating `<projectname_files.cmake>`
    ```cmake
    set(FILES
        ...
        Source/AutoGen/MyFirstNetworkComponent.AutoComponent.xml
    )
    ```

> {{< note >}}Feel free to delete this temporary auto-component, or use it as a starting point, once you're ready to build further multiplayer components. Find out more in general about multiplayer auto-components [here](/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autocomponents/), or follow [this tutorial](/docs/learning-guide/tutorials/multiplayer/first-multiplayer-component/).{{< /note >}}

Reviewer: Maybe "After completing the setup steps in this guide, you can delete the temporary auto-component, or use it.."?

Author: Updated! Also reiterated the fact that at least 1 auto-component must exist in case anyone mistake that sentence as saying you can delete the temp-auto-component without first adding an new one.

> {{< note >}}After completing the setup steps in this guide, you can delete the temporary auto-component and create a new auto-component, or use it as a starting point. There always needs to be at least 1 auto-component. Find out more in general about multiplayer auto-components [here](/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autocomponents/), or follow [this tutorial](/docs/learning-guide/tutorials/multiplayer/first-multiplayer-component/).{{< /note >}}

Reviewer 2: Ditto.

```suggestion
{{< note >}}After completing the setup steps in this guide, you can delete the temporary auto-component and create a new auto-component, or use it as a starting point. There must always be at least one auto-component.

You can learn more about multiplayer auto-components [here](/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autocomponents/), or follow [this tutorial](/docs/learning-guide/tutorials/multiplayer/first-multiplayer-component/).{{< /note >}}
```

## Module and System component setup

In order to use multiplayer functionality, you need to make small changes to the source code to generate descriptors for multiplayer components and then register these components with the Multiplayer Gem.

### Module.cpp changes

Make the following changes to your project's `Code/Source/<ProjectName>Module.cpp`:
1. Include `AutoComponentTypes.h` at the top of the file.
    ```cpp
    #include <Source/AutoGen/AutoComponentTypes.h>
    ```

> 1. Edit the `<ProjectName>Module` constructor to create the component descriptors, allowing Multiplayer components to be registered.

Reviewer 2: ```suggestion
1. Edit the `<ProjectName>Module` constructor to create the component descriptors, which allows Multiplayer components to be registered.
```

    ```cpp
    MultiplayerSampleModule()
        : AZ::Module()
    {
        ...
        CreateComponentDescriptors(m_descriptors); //< Add this line to register your projects multiplayer components
    }
    ```
    {{< important >}}Make sure the call to `CreateComponentDescriptors()` is the *last* line of the constructor.
    {{< /important >}}

### SystemComponent.cpp changes

Make the following changes to your project's `Code/Source/<ProjectName>SystemComponent.cpp` file.
1. Include `AutoComponentTypes.h` at the top of the file.
    ```cpp
    #include <Source/AutoGen/AutoComponentTypes.h>
    ```
2. Register Multiplayer components with the Gem by updating the `Activate()` function.

    ```cpp
    ...
    void <ProjectName>SystemComponent::Activate()
    {
        ...
        RegisterMultiplayerComponents(); //< Register our gems multiplayer components to assign NetComponentIds
    }
    ```

## Rebuild the project

> Configuring and building is always required after editing CMake and C++ files. Use the [Project Manager](/docs/user-guide/project-config/project-manager/) to rebuild; otherwise instructions for configuring and building via commandline-interface (CLI) can be found [here](/docs/user-guide/build/configure-and-build/).

Reviewer 2: ```suggestion
Configuring and building is always required after editing CMake and C++ files. You can use the [Project Manager](/docs/user-guide/project-config/project-manager/) to rebuild, or configure and build via the command-line interface (CLI) can be found [here](/docs/user-guide/build/configure-and-build/).
```
